components/eam/src/chemistry/mozart/mo_gas_phase_chemdr.F90

```diff
@@ -1570,18 +1570,20 @@ subroutine gas_phase_chemdr(lchnk, ncol, imozart, q, &
        do m = 1,pcnst
           n = map2chm( m )
           if ( n > 0 ) then
-            if ( any( cflx(:ncol,m) /= 0._r8 ) ) then
+           do i=1,ncol
+            if ( cflx(i,m) /= 0._r8  ) then
               if ( .not. any( aer_species == n ) .and. adv_mass(n) /= 0._r8 ) then
                 do k = pver, pver-srf_emit_nlayer+1, -1 ! loop from bottom to top
                   ! kg/m2, tracer mass
-                  wrk(:ncol,k) = adv_mass(n)*vmr(:ncol,k,n)/mbar(:ncol,k) &
-                                    *pdeldry(:ncol,k)*rga
-                  tmp(:ncol) = wrk(:ncol,k) + cflx(:ncol,m)*delt/dble(srf_emit_nlayer)
-                  vmr(:ncol,k,n) = tmp(:ncol)*mbar(:ncol,k)/adv_mass(n)/pdeldry(:ncol,k)/rga
+                  wrk(i,k) = adv_mass(n)*vmr(i,k,n)/mbar(i,k) &
```

Comment:
@wlin7 I understand you modification. But I just curious why wrk(:,k) can cause issue. I mean we have a lot fortran code like this

Reply:
@keziming , this is explained in more detail on the s[pecial meeting note page](https://acme-climate.atlassian.net/wiki/spaces/CM/pages/3727753245/2023-03-28+Special+Topics+-+NBFB+in+v3atm). (number 25 of the first table there).  The problem is not with wrk itself. The problem is, for columns with cflx=0, vmr should not change. But in the old array operation format, via wrk, vmr is first converted back and forth. Analytically, the multiplier followed by multiplying the inverse of the multipliers should cancel out. But in floating calculation, the round-off during each arithmetic operation causing the cancellation to be not exact. When threading or pe-layout change, where the cflx=0 columns belong would change, for example, from a chunk with all cflx=0 to a chunk with partially non-zero, the zero cflx column would see NBFB due to the round-off.